pkg/featurechangestopper/featurechangestopper_test.go

```diff
@@ -0,0 +1,87 @@
+package featurechangestopper
+
+import (
+	"context"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	configv1 "github.com/openshift/api/config/v1"
+
+	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
+
+	"k8s.io/client-go/tools/cache"
+)
+
+func TestTechPreviewChangeStopper_syncHandler(t *testing.T) {
+	tests := []struct {
+		name                     string
+		startingTechPreviewState bool
+		featureGate              string
+		expectedShutdownCalled   bool
+	}{
+		{
+			name:                     "default-no-change",
+			startingTechPreviewState: false,
+			featureGate:              "",
+			expectedShutdownCalled:   false,
+		},
+		{
+			name:                     "default-with-change-to-tech-preview",
+			startingTechPreviewState: false,
+			featureGate:              "TechPreviewNoUpgrade",
+			expectedShutdownCalled:   true,
+		},
+		{
+			name:                     "default-with-change-to-other",
+			startingTechPreviewState: false,
+			featureGate:              "AnythingElse",
+			expectedShutdownCalled:   false,
+		},
+		{
+			name:                     "techpreview-to-techpreview",
+			startingTechPreviewState: true,
+			featureGate:              "TechPreviewNoUpgrade",
+			expectedShutdownCalled:   false,
+		},
+		{
+			name:                     "techpreview-to-not-tech-preview", // this isn't allowed today
+			startingTechPreviewState: true,
+			featureGate:              "",
+			expectedShutdownCalled:   true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			actualShutdownCalled := false
+			shutdownFn := func() {
+				actualShutdownCalled = true
+			}
+			indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+			indexer.Add(&configv1.FeatureGate{
+				ObjectMeta: metav1.ObjectMeta{Name: "cluster"},
+				Spec: configv1.FeatureGateSpec{
+					FeatureGateSelection: configv1.FeatureGateSelection{
+						FeatureSet: configv1.FeatureSet(tt.featureGate),
+					},
+				},
+			})
+			lister := configlistersv1.NewFeatureGateLister(indexer)
+
+			c := &TechPreviewChangeStopper{
+				startingTechPreviewState: tt.startingTechPreviewState,
+				featureGateLister:        lister,
+			}
+
+			err := c.syncHandler(context.Background(), shutdownFn)
```

Comment:
Testing this internal helper is convenient, but it seems like it would be about as convenient and also cover a bunch of vendored library code we're relying on if you exercised the public methods with something like:

```golang
c := New(tt.startingTechPreviewState, blaBlahNewFeatureGateInformer(indexer))
ctx, cancel := context.WithCancel(context.Background())
reaped := false
go func() {
  c.Run(ctx, shutdownFn)
  reaped = true
}()
time.Sleep(time.Minute)
cancel()
while !reaped {
  time.Sleep(time.Second)
}
```

Reply:
Based on past experience with controllers and unit tests, I recommend against that.  the time based approach is slow to run, so people don't and then when you tighten it fails in weird ways.  It also requires more plumbing in the general case.

Reply:
I'm fine punting, but may come back to this later on.